spec/curl_context.cr

```diff
@@ -86,3 +68,91 @@ macro with_server(&block)
 
   {{block.body}}
 end
+
+# # Creates a context for specing controllers
+
+# Use context by manually instantiating the entire context with IO::Memory in body and output
+def context(
+  method : String,
+  path : String,
+  headers : HTTP::Headers? = nil,
+  body : String | Bytes | IO | Nil = nil,
+  version = "HTTP/1.1",
+  response_io = IO::Memory.new,
+  **header_opts
+)
+  headers ||= HTTP::Headers.new
+  header_opts.each do |key, value|
+    headers.add(key.to_s.split(/-|_/).map(&.capitalize).join("-"), value.to_s)
+  end
+  response = HTTP::Server::Response.new(response_io, version)
+  request = HTTP::Request.new(method, path, headers, body, version)
+  HTTP::Server::Context.new request, response
+end
+
+# Helper to instantiate context
+private def instantiate_context(method : String, route : String, route_params : Hash(String, String)? = nil, headers : Hash(String, String)? = nil, body : JSON::Any? = nil)
+  headers_io = HTTP::Headers.new
+
+  if !headers.nil?
+    headers.each do |key, value|
+      headers_io.add(key, value)
+    end
+  end
+
+  body_io = body.nil? ? IO::Memory.new : IO::Memory.new(body)
+  ctx = context(method, route, headers: headers_io, body: body_io)
+  ctx.route_params = route_params unless route_params.nil?
+  ctx.response.output = IO::Memory.new
+
+  ctx
+end
+
+record ContextResponse, status_code : Int32, body : String, object : HTTP::Server::Response
+
+# Use context by instantiating the context without specifying body, output IO::Memory
+def context(method : String, route : String, route_params : Hash(String, String)? = nil, headers : Hash(String, String)? = nil, body : JSON::Any? = nil, &block)
+  ctx = instantiate_context(method, route, route_params, headers, body)
+  yield ctx
+  res = ctx.response.not_nil!
+
+  ContextResponse.new(res.status_code, res.output.to_s, res)
+end
+
+# Use context by instantiating the context without specifying body, output IO::Memory, instantiating controller in each block
+module Controller(T)
+  extend self
+
+  def context(method : String, route : String, route_params : Hash(String, String)? = nil, headers : Hash(String, String)? = nil, body : JSON::Any? = nil, &block)
+    ctx = instantiate_context(method, route, route_params, headers, body)
+    instance = T.new(ctx)
+    yield instance
+    res = ctx.response.not_nil!
+
+    ContextResponse.new(res.status_code, res.output.to_s, res)
```

Comment:
I don't think it's worth responding with anything other than the `HTTP::Server::Response`, given that all the other values are available on the response object

Reply:
When we do specs, we only use status_code, and output.to_s 90% of the time. People can access any other metadata information through `ContextResponse#object`.

The only thing meaningful in the HTTP::Server::Response are `@status` and `class Output`, but even the `class Output`, we only want to `.to_s` it.

In addition, if we respond with my option here, we can also deserialise it into a specified class/ struct every single time we do specs, whereas if we only response with HTTP::Server::Response then we still need to `Model.from_json(ctx.response.output.to_s)` every time

Reply:
I don't think that's enough justification to add another layer in, IMO.
- You still have to do `status_code` on the object, as well as the `HTTP::Server::Response`
- You don't necessarily want to `.to_s`, the IO should be enough for a lot of things, like `Object.from_json`

Reply:
I don't think it is possible to do `Object.from_json(HTTP::Server::Response::Output)` (modifying the above example)

```crystal

  it "should spec #index the most verbose way" do
    # Instantiate the controller
    body_io = IO::Memory.new
    ctx = context("GET", BASE, body: body_io, authorization: "X")
    ctx.route_params = {"verbose" => "true"}
    ctx.response.output = IO::Memory.new
    Users.new(ctx).index

    # Expectation
    ctx.response.status_code.should eq 200
    parsed = UserRes.from_json(ctx.response.output) # also does not work if changed to ctx.response.output.as(IO)
    parsed.size.should eq(5)
    parsed.should contain({name: "James", state: "NSW"})
  end

```

Reply:
Stacktrace:
```bash
       Unexpected token: <EOF> at 1:1 (JSON::ParseException)
         from ../../../../../usr/local/Cellar/crystal/0.35.1_1/src/json/pull_parser.cr:528:5 in 'parse_exception'
         from ../../../../../usr/local/Cellar/crystal/0.35.1_1/src/json/pull_parser.cr:524:5 in 'unexpected_token'
         from ../../../../../usr/local/Cellar/crystal/0.35.1_1/src/json/pull_parser.cr:68:7 in 'initialize'
         from ../../../../../usr/local/Cellar/crystal/0.35.1_1/src/json/pull_parser.cr:30:3 in 'new'
         from ../../../../../usr/local/Cellar/crystal/0.35.1_1/src/json/from_json.cr:12:3 in 'from_json'
         from spec/context_spec.cr:15:5 in '->'
         from ../../../../../usr/local/Cellar/crystal/0.35.1_1/src/primitives.cr:255:3 in 'internal_run'
         from ../../../../../usr/local/Cellar/crystal/0.35.1_1/src/spec/example.cr:33:16 in 'run'
         from ../../../../../usr/local/Cellar/crystal/0.35.1_1/src/spec/context.cr:18:23 in 'internal_run'
         from ../../../../../usr/local/Cellar/crystal/0.35.1_1/src/spec/context.cr:330:7 in 'run'
         from ../../../../../usr/local/Cellar/crystal/0.35.1_1/src/spec/context.cr:18:23 in 'internal_run'
         from ../../../../../usr/local/Cellar/crystal/0.35.1_1/src/spec/context.cr:147:7 in 'run'
         from ../../../../../usr/local/Cellar/crystal/0.35.1_1/src/spec/dsl.cr:270:7 in '->'
         from ../../../../../usr/local/Cellar/crystal/0.35.1_1/src/primitives.cr:255:3 in 'run'
         from ../../../../../usr/local/Cellar/crystal/0.35.1_1/src/crystal/main.cr:45:14 in 'main'
         from ../../../../../usr/local/Cellar/crystal/0.35.1_1/src/crystal/main.cr:114:3 in 'main'
```

Reply:
Might need to rewind the IO

Reply:
there is no `HTTP::Server::Response::Output`, only `HTTP::Server::Response#output -> IO`